src/data.rs

```diff
@@ -65,39 +140,124 @@ pub mod lenses {
+                let mut glyphs = GlyphSet_ {
+                    object: Rc::clone(&data.file.object),
+                    resolved: Rc::clone(&data.file.resolved),
+                };
+                f(&mut glyphs)
             }
         }
 
-        impl Lens2<Rc<Ufo_>, Rc<Glyph_>> for Glyph {
-            fn get<V, F: FnOnce(&Rc<Glyph_>) -> V>(&self, data: &Rc<Ufo_>, f: F) -> V {
-                let glyph = data.get_glyph(&self.0).expect("missing glyph in lens2");
-                f(glyph)
+        impl Lens2<GlyphSet_, GlyphPlus> for Glyph {
+            fn get<V, F: FnOnce(&GlyphPlus) -> V>(&self, data: &GlyphSet_, f: F) -> V {
+                let glyph = data
+                    .object
+                    .get_glyph(&self.0)
+                    .expect("missing glyph in lens2");
+                let glyph = GlyphPlus {
+                    glyph: Rc::clone(glyph),
+                    ufo: Rc::clone(&data.object),
+                    resolved: Rc::clone(&data.resolved),
+                };
+                f(&glyph)
             }
 
-            fn with_mut<V, F: FnOnce(&mut Rc<Glyph_>) -> V>(&self, data: &mut Rc<Ufo_>, f: F) -> V {
+            fn with_mut<V, F: FnOnce(&mut GlyphPlus) -> V>(&self, data: &mut GlyphSet_, f: F) -> V {
                 //FIXME: this is creating a new copy and then throwing it away
                 //this is just so that the signatures work for now, we aren't actually doing any
                 //mutating
-                let mut glyph = data
+                let glyph = data
+                    .object
                     .get_glyph(&self.0)
-                    .map(Rc::clone)
                     .expect("missing glyph in lens2");
+                let mut glyph = GlyphPlus {
+                    glyph: Rc::clone(glyph),
+                    ufo: Rc::clone(&data.object),
+                    resolved: Rc::clone(&data.resolved),
+                };
                 f(&mut glyph)
             }
         }
     }
 }
+
+/// Convert this glyph's path from the UFO representation into a `kurbo::BezPath`
+/// (which we know how to draw.)
+pub fn path_for_glyph(glyph: &Glyph) -> BezPath {
+    /// An outline can have multiple contours, which correspond to subpaths
+    fn add_contour(path: &mut BezPath, contour: &Contour) {
+        let mut close: Option<&ContourPoint> = None;
+
+        if contour.points.is_empty() {
+            return;
+        }
+
+        let first = &contour.points[0];
+        path.move_to((first.x as f64, first.y as f64));
+        if first.typ != PointType::Move {
+            close = Some(first);
+        }
+
+        let mut idx = 1;
+        let mut controls = Vec::with_capacity(2);
+
+        let mut add_curve = |to_point: Point, controls: &mut Vec<Point>| {
+            match controls.as_slice() {
+                &[] => path.line_to(to_point),
+                &[a] => path.quad_to(a, to_point),
```

Comment:
Does this handle the case where a quadratic contour is compressed into an on-curve point, followed by 0 < n < (∞ − 1) off-curve points followed by an on-curve point? Or an outline that consists solely of off-curves?

https://stackoverflow.com/questions/20733790/truetype-fonts-glyph-are-made-of-quadratic-bezier-why-do-more-than-one-consecu#20772557

Reply:
If it does handle those cases it's not on purpose! This also doesn't handle the `qcurve`/ 'implied on-curve points' stuff. It's difficult for me to tell with some of these things how important/common they are, versus being unusual corner cases of the spec; my focus right now is less on being compliant with the spec than on implementing enough features that folks can draw with the app.

There's some reasonable thinking to do about how we want to handle quadratics period, in the near-term; I'm not planning on supporting quadratic path editing initially, so we I haven't thought too much about this.

It would also be nice if we had a good set of data to test with. 

Anyway, I appreciate you calling attention to these details, there's a lot of stuff here I'm not very confident of.

Reply:
That's reasonable. I'm pointing at this now because I once tried to retrofit quad handling into TruFont-WX and it was painful, because quad contours can vary in length as opposed to any other contour type and you have to take care not to code assumptions about contour length into the code flow anywhere. And you may have to expand the implied on-curves any time you want to do math on them (intersections, etc.). Very annoying.

Reply:
For quadratic sources, have a look at the DejaVu and Ubuntu fonts. The former is stored as FontForge SFDs, but you can export them to UFO. I think there is even an all-off-curves contour in there somewhere, I think it was the Regular /six.

Reply:
All these observations are very welcome!

A major question continues to be whether we want to support all possible permutations of the ufo format or if we just want to pick some reasonable subset and support that. Without having a better sense of what exists in the world and what users will want from Runebender, this is a tough question for me to ask, but my inclination is to not support weird edge cases if we can avoid it.

Reply:
I sometimes use a font editor to load quadratic sources and look at the conversion. Sometimes, old sources have quadratics in one layer and the original cubics in another. Also, if the editor is to ever grow TrueType hinting capabilities, quadratics have to be supported in all their glory. For a 0.1.0, supporting cubics only is reasonable.